File: util/android/src/main/java/com/dmi/util/TypeConverters.java
package com.dmi.util;

public abstract class TypeConverters {
    @SuppressWarnings("unchecked")
    public static <T> T stringToType(String value, Class<?> type) throws ParseException {
        try {
            if (value == null || String.class.equals(type) || Void.TYPE.equals(type)) {
                return (T) value;
            } else if (Long.class.equals(type) || Long.TYPE.equals(type)) {
                return (T) Long.valueOf(value);
            } else if (Integer.class.equals(type) || Integer.TYPE.equals(type)) {
                return (T) Integer.valueOf(value);
            } else if (Short.class.equals(type) || Short.TYPE.equals(type)) {
                return (T) Short.valueOf(value);
            } else if (Double.class.equals(type) || Double.TYPE.equals(type)) {
                return (T) Double.valueOf(value);
            } else if (Float.class.equals(type) || Float.TYPE.equals(type)) {
                return (T) Float.valueOf(value);
            } else if (Boolean.class.equals(type) || Boolean.TYPE.equals(type)) {
                return (T) Boolean.valueOf(value);
            } else if (Enum.class.isAssignableFrom(type)) {
                return (T) Enum.valueOf((Class) type, value);
            }
        } catch (Exception e) {
            throw new ParseException(e);
        }
        throw new UnsupportedOperationException();
    }

    public static String typeToString(Object value, Class<?> type) {
        try {
            if (value == null || String.class.equals(type) || Void.TYPE.equals(type)) {
                return (String) value;
            } else if (Long.class.equals(type) || Long.TYPE.equals(type)) {
                return String.valueOf((long) value);
            } else if (Integer.class.equals(type) || Integer.TYPE.equals(type)) {
                return String.valueOf((int) value);
            } else if (Short.class.equals(type) || Short.TYPE.equals(type)) {
                return String.valueOf((short) value);
            } else if (Double.class.equals(type) || Double.TYPE.equals(type)) {
                return String.valueOf((double) value);
            } else if (Float.class.equals(type) || Float.TYPE.equals(type)) {
                return String.valueOf((float) value);
            } else if (Boolean.class.equals(type) || Boolean.TYPE.equals(type)) {
                return String.valueOf((boolean) value);
            } else if (Enum.class.isAssignableFrom(type)) {
                return ((Enum) value).name();
            }
        } catch (Exception e) {
            throw new IllegalArgumentException(e);
        }
        throw new UnsupportedOperationException();
    }

    public static class ParseException extends Exception {
        public ParseException(Exception cause) {
            super(cause);
        }
    }
}

File: util/android/src/main/kotlin/com/dmi/util/TypeConverters.kt
package com.dmi.util

object TypeConverters {
    @Suppress("UNCHECKED_CAST")
    @Throws(ParseException::class)
    fun <T> stringToType(value: String?, type: Class<*>): T {
        try {
            if (value == null || String::class.java == type || Void.TYPE == type) {
                return value as T
            } else if (Long::class.java == type || java.lang.Long.TYPE == type) {
                return java.lang.Long.valueOf(value) as T
            } else if (Int::class.java == type || Integer.TYPE == type) {
                return Integer.valueOf(value) as T
            } else if (Short::class.java == type || java.lang.Short.TYPE == type) {
                return java.lang.Short.valueOf(value) as T
            } else if (Double::class.java == type || java.lang.Double.TYPE == type) {
                return java.lang.Double.valueOf(value) as T
            } else if (Float::class.java == type || java.lang.Float.TYPE == type) {
                return java.lang.Float.valueOf(value) as T
            } else if (Boolean::class.java == type || java.lang.Boolean.TYPE == type) {
                return java.lang.Boolean.valueOf(value) as T
            } else if (Enum::class.java.isAssignableFrom(type)) {
                throw UnsupportedOperationException("not implemented. todo: implement")
            }
        } catch (e: Exception) {
            throw ParseException(e)
        }

        throw UnsupportedOperationException()
    }

    fun typeToString(value: Any?, type: Class<*>): String {
        try {
            if (value == null || String::class.java == type || Void.TYPE == type) {
                return value as String
            } else if (Long::class.java == type || java.lang.Long.TYPE == type) {
                return (value as Long).toString()
            } else if (Int::class.java == type || Integer.TYPE == type) {
                return (value as Int).toString()
            } else if (Short::class.java == type || java.lang.Short.TYPE == type) {
                return (value as Short).toString()
            } else if (Double::class.java == type || java.lang.Double.TYPE == type) {
                return (value as Double).toString()
            } else if (Float::class.java == type || java.lang.Float.TYPE == type) {
                return (value as Float).toString()
            } else if (Boolean::class.java == type || java.lang.Boolean.TYPE == type) {
                return (value as Boolean).toString()
            } else if (Enum::class.java.isAssignableFrom(type)) {
                throw UnsupportedOperationException("not implemented. todo: implement")
            }
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }

        throw UnsupportedOperationException()
    }

    class ParseException(cause: Exception) : Exception(cause)
}
